random/imx6ull-rngb.c

```diff
@@ -0,0 +1,305 @@
+}
+
+
+static int hardwareInit(void)
+{
+	int err;
+
+	/* Reset and wait until rngb is sleeping */
+	*(common_rngb.base + rng_cmd) = 0x40;
+	do {
+		err = *(common_rngb.base + rng_sr) & 0x4;
+	} while (err == 0);
+
+	interrupt(RNGB_IRQ, rngb_intr, NULL, common_rngb.cond, NULL);
+
+	/* Run self test */
+	*(common_rngb.base + rng_cmd) = 0x1;
+	mutexLock(common_rngb.lock);
+	while (common_rngb.intr_st == 0) {
+		condWait(common_rngb.cond, common_rngb.lock, 3000);
+	}
+	mutexUnlock(common_rngb.lock);
+
+	if ((common_rngb.intr_st & ERROR_OCCURRED) != 0) {
+		printf("rngb: self-test failed\n");
+
+		systemCleanup(4);
+		return -1;
+	}
+
+	common_rngb.intr_st = 0;
+
+	/* Run seeding */
+	*(common_rngb.base + rng_cmd) = 0x2;
+	mutexLock(common_rngb.lock);
+	while (common_rngb.intr_st == 0) {
+		condWait(common_rngb.cond, common_rngb.lock, 3000);
+	}
+	mutexUnlock(common_rngb.lock);
+
+	if ((common_rngb.intr_st & ERROR_OCCURRED) != 0) {
+		printf("rngb: error after generating seed\n");
+
+		systemCleanup(4);
+		return -1;
+	}
+
+	common_rngb.intr_st = 0;
+
+	/* Enable automatic seeding */
+	*(common_rngb.base + rng_cr) = 0x10;
+
+	printf("imx6ull-rngb: initialized\n");
+
+	return 0;
+}
+
+
+static int readRandoms(char *buff, size_t size, unsigned mode)
+{
+	union {
+		uint32_t val32;
+		char val8[sizeof(uint32_t)];
+	} reg;
+	unsigned int i;
+	int status;
+
+	if (size == 0) {
+		return 0;
+	}
+
+	mutexLock(common_rngb.lock);
+	i = 0;
+	while (i < size) {
+		if ((i % sizeof(uint32_t)) == 0) {
+			status = *(common_rngb.base + rng_sr);
+			if (((status >> 16) & 1U) == 1) {
+				mutexUnlock(common_rngb.lock);
+				return -EIO;
+			}
+			if (((status >> 8) & 0xf) == 0) {
+				if ((mode & O_NONBLOCK) != 0) {
+					break;
+				}
+				/* wait for new values in FIFO */
+				continue;
+			}
+			reg.val32 = *(common_rngb.base + rng_out);
+		}
+		buff[i] = reg.val8[i % sizeof(uint32_t)];
+		i++;
+	}
+	mutexUnlock(common_rngb.lock);
+
+	return i == 0 ? -EAGAIN : i;
+}
+
+
+static void handleMsg(void *arg)
+{
+	msg_t msg;
+	unsigned long int rid;
+	int err;
+
+	while (1) {
+		err = msgRecv(common_rngb.port, &msg, &rid);
+		if (err < 0) {
+			if (err == -EINTR) {
+				continue;
+			}
+			else {
+				printf("rngb: msgRecv returned error: %s\n", strerror(-err));
+				break;
+			}
+		}
+
+		switch (msg.type) {
+			case mtClose:
+			case mtOpen:
+				msg.o.io.err = msg.i.openclose.oid.id != 0 ? -ENOENT : EOK;
+				break;
+
+			case mtRead:
+				if (msg.o.data != NULL) {
+					msg.o.io.err = readRandoms(msg.o.data, msg.o.size > 512 ? 512 : msg.o.size, msg.i.io.mode);
+				}
+				else {
+					msg.o.io.err = -EINVAL;
+				}
+				break;
+
+			default:
+				msg.o.io.err = -EINVAL;
+				break;
+		}
+
+		msgRespond(common_rngb.port, &msg, rid);
+	}
+}
+
+
+int main(int argc, char **argv)
+{
+	oid_t root;
+	(void)argc;
+	(void)argv;
+
+	while (lookup("/", NULL, &root) < 0) {
+		usleep(10000);
+	}
+
+	if (systemInit() || hardwareInit()) {
+		return EIO;
+	}
+
+	handleMsg(NULL);
+
```

Comment:
System cleanup is missing, any required resources freeing - if `handleMsg()` returns on `msgRecv()` error, also return `0` is misleading if `handleMsg()` terminates.